CIP-output-tagging/README.md

```diff
@@ -0,0 +1,181 @@
+
+We already have redeemers associated with each script to provide the smart contract with information that is only relevant during execution for similar reasons (ie redeemers allow us to associate arbitrary data with a script). This would be to TxOuts what Redeemers are for scripts, allowing us to associate arbitrary data (relevant only during Phase 2 validation) to outputs.
+
+## Specification
+<!-- The technical specification should describe the proposed improvement in sufficient technical detail. In particular, it should provide enough information that an implementation can be performed solely on the basis of the design in the CIP. This is necessary to facilitate multiple, interoperable implementations. This must include how the CIP should be versioned. If a proposal defines structure of on-chain data it must include a CDDL schema in it's specification.-->
+We extend transaction witness set with a new list of arbitrary data associated with transaction outputs (`output_tags`).
+
+### Script context
+
+Scripts are passed information about transactions via the script context.
+We propose to introduce a new field `tag` to the `TxOut`s in the script context.
+
+```haskell
+tag :: Maybe BuiltinData 
+```
+
+Changing the script context will require a new Plutus language version in the ledger to support the new interface. We propose the addition of a `tag` field to all the `TxOut` elements in `TxInfoOutputs` in the script context. This field will contain arbitrary data associated to the output that is relevant to Plutus validators during Phase 2 validation. 
+
+The interface for old versions of the language will not be changed.
+Scripts with old versions cannot be spent in transactions that include output tags, attempting to do so will be a phase 1 transaction validation failure.
+
+Note that for simplicity, we propose that the `tag` field in `TxOut`s in transaction inputs in the script context always be empty (ie set to `Nothing`). There are certainly use-cases where one might want to associate arbitrary data to inputs for use in Phase 2 validation; however these use-cases are few and far between compared to those for outputs. 
+
+
+
+### Output tags in transaction witness set 
+
+In keeping consistent with the design of redeemers, we propose to introduce this information into the witness set as redeemers are.
+
+### CDDL
+
+The CDDL for the transaction witness set will change as follows to reflect the new field.
+```
+transaction_witness_set =
+  { ? 0: [* vkeywitness ]
+  , ? 1: [* native_script ]
+  , ? 2: [* bootstrap_witness ]
+  , ? 3: [* plutus_v1_script ]
+  , ? 4: [* plutus_data ]
+  , ? 5: [* redeemer ]
+  , ? 6: [* plutus_v2_script ]
+  , ? 7: [* plutus_v3_script ]
+  , ? 8: [* output_tag ] ; 
+  }
+
+output_tag = [ index: unit, data: plutus_data ]
```

Comment:
also modifying the redeemer indirectly causes the body to be invalid, because redeemers and datums in the witness set are used to calculate the `scriptDataHash`

modifying these will result in a different `scriptDataHash`, hence the body is invalid

Reply:
> The user signs the transaction body, not the witness set.

The user does sign some parts of the witness set (redeemers and datums)  indirectly through through the script hash integrity check.

Reply:
Yes that is correct I corrected myself in the coment

Reply:
Right the idea would be for the user to sign this information the exact same way they sign redeemers in datums (ie adjust  script hash integrity check accordingly). I will update this CIP to reflect that.

Reply:
> the idea would be for the user to sign this information the exact same way they sign redeemers in datums (ie adjust script hash integrity check accordingly

@colll78 please don't
 
`scriptDataHash` is already a pain, I would suggest instead to move the field as is inside the body

Reply:
Maybe also move the redeemer in the body in the process?

Reply:
I agree that it would be best if redeemers were directly in the transaction body and I would fully support a separate CIP to that end, however, given that currently they are signed via `scriptDataHash` I don't want to introduce inconsistency by moving these into the body in which case tooling would have to accommodate them differently than they do normal redeemers. The idea is to keep them as close from an implementation perspective to transaction redeemers to lower the implementation burden.

Reply:
@colll78 trust me, `scriptDataHash` is much worse than a small inconsistency

Reply:
Should we not then create a CIP to move redeemers into the tx body?